### ISSUE TITLE
JmDNS sample codes

### ISSUE THREAD

**Issue by janossch** [jmdns/jmdns#319]
In my opinion codes under `src/sample/java` are difficult to maintain. Since they are located outside of the maven module they are not even built during a regular build, so that it is easy to break.

One possible solution could be to move all sample codes under `src/test/sample` since ultimately they are just _"test"_ codes.

What do you think?

**Comment by ViToni** [comment]
IMO it would be prefererable to change the build script to build the sample code e.g. as part of the verify step.

**Comment by janossch** [comment]
Another benefit of moving the files under `src/test/sample` would be to get an out of the box IDE support for them, without having to move or copy+paste them somewhere or do any manual steps to work with them. They won't even packaged into the artifact jar by default.

@ViToni I'm not sure what you meant on "verify step". Do you mean maven [verify](https://maven.apache.org/guides/introduction/introduction-to-the-lifecycle.html#a-build-lifecycle-is-made-up-of-phases) phase?

**Comment by ViToni** [comment]
> Another benefit of moving the files under `src/test/sample` would be to get an out of the box IDE support for them, without having to move or copy+paste them somewhere or do any manual steps to work with them. They won't even packaged into the artifact jar by default.
> 
Me, personally I wouldn't like these files to go to `test`. Maybe the project could be changed to multi-module, which would cover the IDE aspect and the "not into `test`" point...

> @ViToni I'm not sure what you meant on "verify step". Do you mean maven [verify](https://maven.apache.org/guides/introduction/introduction-to-the-lifecycle.html#a-build-lifecycle-is-made-up-of-phases) phase?

Exactly that.

**Comment by kaikreuzer** [comment]
I'd agree with @ViToni, `test` does not sound like a good place. Note that [this sister project](https://github.com/jupnp/jupnp) is also multi-module and has a `tools` part where such things like the browser app are contained.

**Comment by janossch** [comment]
As far as I can dig out from the history these files was under `src/main/java` in the past: https://github.com/jmdns/jmdns/commit/7ded59c63d5d1303c570a5ef4ce841ef9fccd5a3

Additionally the latest version of the `pom.xml` has some instructions to not export samples, but in reality they aren't even there:
https://github.com/jmdns/jmdns/blob/26e30a5e31593d8a9dbc0216b45c106638ba2d19/pom.xml#L164

What do you think if we move them back (maybe under `javax.jmdns.tools`) and exclude them from the final jar (as it was previously)?
Or would the multi-module project structure be the way to go?

**Comment by ViToni** [comment]
IMO multi-module is the way to go.

**Comment by ViToni** [comment]
> As far as I can dig out from the history these files was under `src/main/java` in the past: [7ded59c](https://github.com/jmdns/jmdns/commit/7ded59c63d5d1303c570a5ef4ce841ef9fccd5a3)
> 
> Additionally the latest version of the `pom.xml` has some instructions to not export samples, but in reality they aren't even there:
> 
> https://github.com/jmdns/jmdns/blob/26e30a5e31593d8a9dbc0216b45c106638ba2d19/pom.xml#L164
> 
These instructions just affect the manifest and are useful for OSGi. The occurrence of the packages ˋcom.strangeberry.*,samples` might be a leftover from the time before the files were moved.
It doesn't affect the actual inclusion of the files.

**Comment by janossch** [comment]
Yeah that's so far so good. Obviously that is just remnants, once upon a time someone wanted to hide these packages in OSGi environments.

But similarly what if we completely remove (do not include) the compiled classes in the jar (when moving them under `src/main/java`).

I believe that can be achieved by specifying  an `exlude` in `maven-jar-plugin`

https://maven.apache.org/plugins/maven-jar-plugin/examples/include-exclude.html

I don't think (but maybe I'm wrong) that these samples and tools will change/expand so frequent which worth a separate module.

**Comment by ViToni** [comment]
Multi-module would be still the way to go IMHO as it would compile code and verify it's still valid and separate it sample code from library code.
